Goto analyzer

danpoe opened the pull request:
I'm taking over the goto analyzer branch/PR (#424) from @martin-cs which is now replaced by this PR. It contains everything from #424 plus whitespace fixes and some fixes to the constant propagator.

tautschnig commented:
Would you be able to do some merging of commits (there are multiple commits with the same commit message) and possibly also some tweaks to the commit messages? A few of them could be phrased in a more neutral tone.

danpoe commented:
Ok, I've tweaked the commit messages. The four constant propagator regression tests are not expected to work at present and I've marked them as FUTURE.

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -339,11 +460,26 @@ void constant_propagator_domaint::valuest::output(
+      ++it)
+  {
+    out << ' ' << it->first << "=" << from_expr(ns, "", it->second) << '\n';
+  }
```
I believe this reverts an intended change.

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -380,58 +516,75 @@ Function: constant_propagator_domaint::valuest::merge
+  // dummy
+  const symbol_tablet symbol_table;
+  const namespacet ns(symbol_table);
+
```
Probably I'm being dense, but I just don't see where this is being used?

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -380,58 +516,75 @@ Function: constant_propagator_domaint::valuest::merge
   }
 
-  bool changed = false;
+  assert(!is_bottom && !src.is_bottom);
```
`!is_bottom` seems rather trivially true?

tautschnig replied:
You might still want to consider this one?

danpoe replied:
I'd like to keep this one. It's rather trivial but it forms a precondition for the code block below. That helps me as I change the code or move things around.

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -380,58 +516,75 @@ Function: constant_propagator_domaint::valuest::merge
+
+    set_to_top();
+    assert(expr_map.empty());
+    assert(!is_bottom);
```
I think we've established `!is_bottom` by now ... ;-)

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -454,24 +607,27 @@ bool constant_propagator_domaint::valuest::meet(const valuest &src)
 
   bool changed = false;
```
Not in your commit, but there is some stray whitespace here.

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -454,24 +607,27 @@ bool constant_propagator_domaint::valuest::meet(const valuest &src)
-      set_to(src_replace_pair.first, src_replace_pair.second);
-      changed=true;
+      set_to(it->first, it->second);
+      changed = true;
```
All of the above changes appear to revert some intentional changes.

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -495,7 +651,31 @@ bool constant_propagator_domaint::merge(
 {
-  return values.merge(other.values);
+  const symbol_tablet symbol_table;
+  const namespacet ns(symbol_table);
```
Those lines should probably be moved inside the `#if 0` (and then one wonders what the secrets here are).

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -551,23 +731,26 @@ void constant_propagator_ait::replace(
+      exprt &rhs=to_code_assign(it->code).rhs();
       s_it->second.values.replace_const(rhs);
-      rhs = simplify_expr(rhs, ns);
+      rhs=simplify_expr(rhs, ns);
```
Could use `simplify(rhs, ns);`

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -551,23 +731,26 @@ void constant_propagator_ait::replace(
         s_it->second.values.replace_const(*o_it);
-        *o_it = simplify_expr(*o_it, ns);
+        *o_it=simplify_expr(*o_it, ns);
       }
```
Would you mind replacing this loop by
```
for(auto &op : args)
{
  s_it->second.values.replace_const(op);
  simplify(op, ns);
}
```

tautschnig commented on src/analyses/constant_propagator.h:
```diff
@@ -9,103 +9,152 @@ Author: Peter Schrammel
 #ifndef CPROVER_ANALYSES_CONSTANT_PROPAGATOR_H
 #define CPROVER_ANALYSES_CONSTANT_PROPAGATOR_H
 
+#include <iostream>
```
Use `#include <iosfwd>`

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -331,6 +333,54 @@ void dep_graph_domaint::output(
+
+\*******************************************************************/
+
+
```
A single blank line suffices

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -331,6 +333,54 @@ void dep_graph_domaint::output(
+
+  for(dep_graph_domaint::depst::const_iterator cdi=control_deps.begin();
+      cdi!=control_deps.end();
+      ++cdi)
```
`for(const auto &cd : control_deps)`

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -331,6 +333,54 @@ void dep_graph_domaint::output(
+
+  for(dep_graph_domaint::depst::const_iterator ddi=data_deps.begin();
+      ddi!=data_deps.end();
+      ++ddi)
```
`for(const auto &dd : data_deps)`

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -331,6 +333,54 @@ void dep_graph_domaint::output(
+}
+
+
+
```
Just a single blank line

tautschnig commented on src/analyses/interval_domain.cpp:
```diff
@@ -498,3 +498,65 @@ exprt interval_domaint::make_expression(const symbol_exprt &src) const
+
+ Outputs: A simplified version of the expression.
+
+ Purpose: Uses the domain to simplify a given expression using context-specific information.
```
Please break this long line

tautschnig commented on src/analyses/interval_domain.cpp:
```diff
@@ -498,3 +498,65 @@ exprt interval_domaint::make_expression(const symbol_exprt &src) const
+    {
+      exprt e;
+      e.make_true();
+      return e;
```
`return true_exprt();`

tautschnig commented on src/analyses/interval_domain.cpp:
```diff
@@ -498,3 +498,65 @@ exprt interval_domaint::make_expression(const symbol_exprt &src) const
+    {
+      exprt e;
+      e.make_true();
+      return e;
```
`return true_exprt();`

tautschnig commented on src/analyses/replace_symbol_ext.cpp:
```diff
@@ -38,20 +38,7 @@ bool replace_symbol_extt::replace(exprt &dest) const
-    }
-  }
-  else if(dest.id()==ID_symbol)
+  if(dest.id()==ID_symbol)
```
With this change, replace_symbol_extt is no longer any different from replace_symbolt!?

danpoe replied:
Ah right. ``simplify()`` does not simplify address of expressions with a constant as the object anyways, so it does not matter if we replace the object in address of expressions with a constant. I've now removed ``replace_symbol_extt``.

danpoe replied:
Or probably that was a bad idea. For ``domain_simplify()`` we still need ``replace_symbol_extt``.

tautschnig replied:
Maybe it would just be much better to have an option with replace_symbol to state whether addresses should be replaced? Duplicate code is so hard to maintain...

danpoe replied:
Or I think we can just generally don't replace symbols in address of expressions with constant expressions. It seems that this would never be meaningful.

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -185,128 +291,188 @@ Function: goto_analyzer_parse_optionst::doit
-  if(goto_model(cmdline.args))
-    return 6;
+    goto_functionst::function_mapt::const_iterator f_it
+      =goto_model.goto_functions.function_map.find(
```
"=" belongs to the preceding line.

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -185,128 +291,188 @@ Function: goto_analyzer_parse_optionst::doit
+        if(!ofs)
+        {
+          error() << "Failed to open json output `"
+          << json_file << "'" << eom;
```
Indent of "<<" should match the above "<<"

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -185,128 +291,188 @@ Function: goto_analyzer_parse_optionst::doit
+    // Run the analysis
+    bool result=true;
+    if(options.get_bool_option("show"))
+      result=static_show_domain(
```
Here and below: break the line after "="

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -185,128 +291,188 @@ Function: goto_analyzer_parse_optionst::doit
-      local_may_aliast local_may_alias(it->second);
-      local_may_alias.output(std::cout, it->second, ns);
-      std::cout << '\n';
+      error()<<"No task given"<<eom;
```
Whitespace around "<<", please.

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -328,19 +494,16 @@ bool goto_analyzer_parse_optionst::set_properties()
   {
     error() << e << eom;
     return true;
   }
```
Here and below: don't we now have multiple, nested exception handlers?

danpoe replied:
Hm?

tautschnig replied:
Ah, sorry, please ignore my comment. It seems we don't have an exception handler in `doit`.

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -462,7 +622,7 @@ void goto_analyzer_parse_optionst::help()
   std::cout <<
     "\n"
-    "* * GOTO-ANALYSER " CBMC_VERSION " - Copyright (C) 2016 ";
+    "* * GOTO-ANALYZER " CBMC_VERSION " - Copyright (C) 2016 ";
```
It may be worth bumping the year.

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -477,16 +637,30 @@ void goto_analyzer_parse_optionst::help()
-    " --non-null                   non-null analysis\n"
+    "Abstract interpreter options:\n"
+    " --flow-sensitive             use flow-sensitive abstract interpreter\n"
+    " --concurrent                 use concurrent abstract interpreter\n"
```
I think this is misleading: it's a concurrency-aware AI, not a concurrently-executing one.

tautschnig commented on src/goto-analyzer/static_analyzer.cpp:
```diff
@@ -8,25 +8,28 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "static_analyzer.h"
 
+template<class analyzerT>
```
Use `typename` instead of `class` (so as to permitting passing in, e.g., a struct or typedef.

tautschnig commented on src/goto-analyzer/static_analyzer.cpp:
```diff
@@ -124,57 +109,62 @@ void static_analyzert::plain_text_report()
   }
 
-  status() << "SUMMARY: " << pass << " pass, " << fail << " fail, "
+  status() << "SUMMARY: " << pass << " pass, " << fail << " fail if reachable, "
```
If this is intended as user-visible status info, then it shouldn't be all-caps (see above).

tautschnig commented on src/goto-analyzer/static_show_domain.cpp:
```diff
@@ -0,0 +1,99 @@
+  message_handlert &message_handler,
+  std::ostream &out)
+{
+  ai_baset *domain = NULL;
```
No whitespace; and: would a `std::unique_ptr` be helpful?

tautschnig commented on src/goto-analyzer/static_show_domain.h:
```diff
@@ -0,0 +1,26 @@
+
+#include <util/message.h>
+#include <util/namespace.h>
+#include <util/options.h>
```
See elsewhere: I'd opt for forward declarations; though namespace.h certainly is redundant.

tautschnig commented on src/goto-analyzer/static_simplifier.cpp:
```diff
@@ -0,0 +1,303 @@
+  exprt simplified=domain[i_it].domain_simplify(i_it->guard, ns);
+  unsigned return_value=(simplified==i_it->guard)?0:1;
+  i_it->guard=simplified;
+  return return_value;
```
Multiple notes:
1. The return value is apparently Boolean.
2. Comparisons of expressions are costly. Instead, I'd suggest that domain_simplify has a bool return type to report whether simplification was performed (my ealier suggestions of using `simplify` instead of `simplify_expr` will help in this regard.
3. Do not replace `i_it->guard` when it hasn't even changed.

tautschnig commented on src/goto-analyzer/static_simplifier.cpp:
```diff
@@ -0,0 +1,303 @@
+  };
+
+  counterst simplified={0, 0, 0, 0, 0};
+  counterst unmodified={0, 0, 0, 0, 0};
```
This is what constructors are meant for (RAII).

tautschnig commented on src/goto-analyzer/static_simplifier.cpp:
```diff
@@ -0,0 +1,303 @@
+      {
+        unsigned result=simplify_guard(i_it);
+        simplified.asserts+=result;
+        unmodified.asserts+=(1-result);
```
To close the loop on the return-type-is-bool bit:
```
if(!simplify_guard(i_it))
  ++simplified.asserts;
else
  ++unmodified.asserts;
```

tautschnig commented on src/goto-analyzer/static_simplifier.cpp:
```diff
@@ -0,0 +1,303 @@
+
+        assign.lhs()=simp_lhs;
+        assign.rhs()=simp_rhs;
+        i_it->code=assign;
```
To be honest, there are too many redundant expression copies in this branch. How about making "assign" a non-const reference, then that last line could go away. (And the first one would not be a copy.)

tautschnig commented on src/goto-analyzer/static_simplifier.cpp:
```diff
@@ -0,0 +1,303 @@
+        code_function_callt fcall(to_code_function_call(i_it->code));
+
+        exprt new_function=domain[i_it].domain_simplify(fcall.function(), ns);
+        result|=(new_function==fcall.function()) ? 0 : 1;
```
There is no need for `|=` here.

tautschnig commented on src/goto-analyzer/static_simplifier.cpp:
```diff
@@ -0,0 +1,303 @@
+
+        for(exprt::operandst::iterator o_it=args.begin();
+            o_it!=args.end();
+            ++o_it)
```
`for(auto &op : args)` and the |= below is questionable.

tautschnig commented on src/goto-analyzer/static_simplifier.cpp:
```diff
@@ -0,0 +1,303 @@
+          *o_it=new_arg;
+        }
+
+        i_it->code=fcall;
```
See above: non-const references will do, no need to copy back and forth.

tautschnig commented on src/goto-analyzer/static_simplifier.h:
```diff
@@ -0,0 +1,26 @@
+#include <iosfwd>
+
+#include <util/message.h>
+#include <util/namespace.h>
```
See above: this isn't need, and forward declarations are (personally) preferred.

tautschnig commented on src/goto-programs/remove_unreachable.cpp:
```diff
@@ -57,3 +57,10 @@ void remove_unreachable(goto_programt &goto_program)
+  Forall_goto_functions(f_it, goto_functions)
+    remove_unreachable(f_it->second.body);
+}
+
```
No blank line at EOF, please.

danpoe commented on src/analyses/ai.cpp:
```diff
@@ -82,6 +82,158 @@ void ai_baset::output(
+
+ Outputs: The JSON object
+
+ Purpose: Output the domains for the whole program as JSON
```
I think there's not much overlap. @thk123's code is concerned with converting the goto program to json while here we're converting abstract states.

tautschnig replied:
Ok!

danpoe commented:
Thanks for the detailed review! I think I've addressed all the comments.

tautschnig commented on regression/Makefile:
```diff
@@ -3,3 +3,11 @@ DIRS = ansi-c cbmc cpp goto-instrument goto-analyzer cbmc-java
+			$(MAKE) -C "$$dir" clean; \
+		fi; \
+	done;
+
```
No blank lines at end-of-file, please. (Applies to regression/**/Makefile)

danpoe replied:
Fixed.

tautschnig replied:
There's still a blank line here, isn't it?

tautschnig commented on regression/goto-analyzer/intervals_02/main.c:
```diff
@@ -2,10 +2,8 @@
   int x;
-  if (x > 0) {
-    if (x < 20) {
+  if (x > 0 && x < 20) {
```
I might be missing something, but what's the rationale for this change?

danpoe replied:
I think this change was introduced by @martin-cs.

tautschnig replied:
@martin-cs  ?

tautschnig commented on src/analyses/ai.h:
```diff
@@ -69,6 +81,28 @@ class ai_domain_baset
+  {
+    return condition;
+  }
+#endif
```
You might just remove this code.

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -167,8 +237,10 @@ bool constant_propagator_domaint::two_way_propagate_rec(
 
+  bool change=false;
+
+#if 0
```
It seems you are disabling the entire procedure. Would you mind adding a comment?

danpoe replied:
This seems to be buggy at present. I plan to re-enable it in a future PR.

martin-cs commented:
> I think there's not much overlap. @thk123's code is concerned with converting the goto program to json while here we're converting abstract states.
We should use @thk123's conversion of an instruction to JSON as at the moment we output it as a (structured) string.

danpoe commented:
I suggest that whoever's pull request (of @thk123 or me) is merged not first changes the code in ``ai.cpp`` to output the goto instruction using @thk123 new functionality.

martin-cs commented:
On Mon, 2017-02-06 at 04:26 -0800, danpoe wrote:
> I suggest that whoever's pull request (of @thk123 or me) is merged not first changes the code in ``ai.cpp`` to output the goto instruction using @thk123 new functionality.
Yes.

martin-cs commented:
> > @@ -2,10 +2,8 @@
>  
>  int main(){
>    int x;
> -  if (x > 0) {
> -    if (x < 20) {
> +  if (x > 0 && x < 20) {
> 
> I might be missing something, but what's the rationale for this change?

At one point the handling of 'assume' (for path sensitivity) was
"if(assumption.id() == ID_le) ...' which would then fail to handle the
conjunction.  This was a change Lucas and I discussed and I think he
implemented.  I'm not sure of the relationship between the test and this
but ... I don't think they are entirely unconnected.


> > @@ -69,6 +81,28 @@ class ai_domain_baset
>    //
>    // This computes the join between "this" and "b".
>    // Return true if "this" has changed.
> +
> +  // This method allows an expression to be simplified / evaluated using the
> +  // current value of the domain.  It is used to evaluate assertions and in
> +  // program simplification
> +#if 0
> +  virtual exprt domain_simplify(
> +    const exprt &condition,
> +    const namespacet &ns,
> +    const bool lhs=false) const
> +  {
> +    return condition;
> +  }
> +#endif
> 
> You might just remove this code.

Please don't.  This is one of the changes to the interface of abstract
domains that I think we really need; hence I added it.  It allows path
sensitivity and a number of other "evaluate this expression" pieces of
functionality to be implemented very easily, including the --simplify
and --verify modes.  It also can have a trivial default implementation
as it is backwards compatible.

tautschnig commented:
With regard to @martin-cs last comment: can this note please just go in the source code together with the `#if 0` - that sort of information will really help, moving forward. Thanks!

danpoe commented:
I think this is ready to be merged!

The instructions in the json abstract state output are still represented as strings using ``goto_program.output_instruction()`` as the current goto functions json output does not have a separate ``goto_program.output_json_instruction()``. However, I don't think this is a blocker for this being merged.

danpoe commented:
I've added now also ``is_bottom()`` and ``is_top()`` to ``ai_domain_baset``. ``is_bottom()`` is used to avoid to apply ``transform()`` on a state that is bottom when doing the edge from the function exit to the function return site, and it will also be needed by a future PR about loop bound computation.

forejtv commented:
@danpoe, can you please rebase this to see if all tests pass?

forejtv commented:
Travis tests are failing. @danpoe, can you have a look at what the problem is?

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -245,6 +247,7 @@ bool constant_propagator_domaint::two_way_propagate_rec(
 
   bool change=false;
 
+#if 0
```
Would you mind explaining *why* you are disabling it?

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -93,7 +93,9 @@ void constant_propagator_domaint::transform(
-  assert(!values.is_bottom);
+  //assert(!values.is_bottom);
+  if(values.is_bottom)
+    return;
```
As far as I'm aware this reverts a change introduced in this commit series - would you mind squashing these changes?

tautschnig commented on regression/goto-instrument-wmm-core/Makefile:
```diff
@@ -21,7 +21,7 @@ tests.log: ../test.pl
 	@for dir in *; do \
                 if [ -d "$$dir" ]; then \
-                        rm $$dir/*.txt $$dir/*.dot $$dir/*.gb $$dir/*.out; \
+                        rm -f $$dir/*.txt $$dir/*.dot $$dir/*.gb $$dir/*.out; \
```
It may be worthwhile going straight for `$(RM)` (possibly everywhere) as this will be appropriately defined on each platform.

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -76,16 +70,33 @@ void constant_propagator_domaint::transform(
+  bool have_dirty=false;
+  bool ignore_unresolved_calls=false;
+
+  if(cp!=nullptr)
```
Shouldn't this even be asserted? I'd be surprised if that dynamic cast failed, but maybe I'm missing something.

danpoe replied:
This is so the constant domain also works when used as a template parameter for ``ait``, which is how it is used in a different branch (loop-bounds).

tautschnig replied:
Ok! It may be worthwhile adding comments about such bits of knowledge that only the developer has.

danpoe replied:
Ok, I've added a comment above.

tautschnig commented on src/analyses/constant_propagator.cpp:
```diff
@@ -167,8 +253,11 @@ bool constant_propagator_domaint::two_way_propagate_rec(
 
+  bool change=false;
+
+  // this seems to be buggy at present
```
How has that been determined?

danpoe replied:
We got some wrong results on an industrial example which went away when two-way-propagation was disabled. I'll look into it more and try to provide a regressiont test and fix.

tautschnig commented on src/analyses/constant_propagator.h:
```diff
@@ -9,87 +9,138 @@ Author: Peter Schrammel
-    void set_to_bottom()
+    // set whole state
+
+    inline void set_to_bottom()
```
No need for `inline` in the class declaration (several).

tautschnig commented on src/analyses/constant_propagator.h:
```diff
@@ -100,23 +151,40 @@ class constant_propagator_domaint:public ai_domain_baset
+    const namespacet &ns) : dirty(goto_functions)
   {
+    assert(false);
+
```
Why are those entry points disallowed?

danpoe replied:
They are currently unused. I think we can remove them.

tautschnig replied:
Then just remove them to avoid dead code idling around.

tautschnig commented on src/analyses/constant_propagator.h:
```diff
@@ -100,23 +151,40 @@ class constant_propagator_domaint:public ai_domain_baset
+    const namespacet &ns) : dirty(goto_function)
   {
+    assert(false);
+
```
Why are those entry points disallowed?

tautschnig commented on src/util/replace_symbol.h:
```diff
@@ -56,6 +59,11 @@ class replace_symbolt
     type_map.clear();
   }
 
+  inline bool empty() const
```
Omit `inline`

thk123 commented:
@danpoe The compile errors are caused by [this function](https://github.com/diffblue/cbmc/pull/472/files#diff-809169cc517e7f7f18c974833ad9494dR32) being marked as `override final` but not overriding anything (maybe because of the type?):

```C++
bool merge(
  const constant_propagator_domaint &other,
  locationt from,
  locationt to) final override;
```

I "fixed" this in this commit to get it building on my branch: https://github.com/danpoe/cbmc/pull/1/commits/3a8d15462dcac9cbd16c2436de4c9e5ed3d5b566

martin-cs commented:
> >  
> +  bool change=false;
> +
> +  // this seems to be buggy at present
> 
> We got some wrong results on an industrial example which went away when two-way-propagation was disabled. I'll look into it more and try to provide a regressiont test and fix.

The variable_sensitivity_domaint may obsolete this code (by, correctly
handling pointers and the like) and so is a little of a higher priority
than a detailed fix here.

martin-cs commented:
>  int main(){
>    int x;
> -  if (x > 0) {
> -    if (x < 20) {
> +  if (x > 0 && x < 20) {
> 
> @martin-cs  ?

See further down the thread; github doesn't seem to handle e-mail
responses to specific issues.

tautschnig commented on src/analyses/ai.cpp:
```diff
@@ -18,6 +19,51 @@ Author: Daniel Kroening, kroening@kroening.com
+{
+  std::ostringstream out;
+  output(out, ai, ns);
+  xmlt xml("abstract_state");
```
Do XML vs JSON use different rules for CaMeL_caseing? Just wondering.

peterschrammel replied:
Json follows Google rules. XML follows the convention that Daniel started with.

tautschnig replied:
Ok, thanks!

thk123 commented:
@forejtv You said 

> that error should not happen anymore, let me investigate

For the windows build, is there some way to trigger a rebuild on this (or perhaps @danpoe you could rebase to force a rebuild?)

danpoe commented:
@thk123: Ok, I rebased again.

tautschnig commented:
@danpoe It seems 6 goto-analyzer tests are failing as the expected output isn't present (the stats don't match).

danpoe commented:
It's strange that this only happens for the first build. From looking at the output of the tests it may be that assertions are translated differently for this build (as a higher number of gotos in the code is reported, and for some outcomes for which we expect ``unknown`` we get ``failure (if reachable)``). Any idea why this happens just for this build?

tautschnig commented:
Are we actually running tests in all (later) builds?

danpoe commented:
Yes, we're running the tests for all builds. I've now commented out the include of ``assert.h`` for the failing tests to see what happens. In any case, the result of goto-analyzer was sound for the failing tests but different than expected.

danpoe commented:
I've removed the include of ``assert.h`` now for the goto-analyzer regression tests which gives the same behavior also for the first Travis build. Additionally I've added two new tests that do use ``assert.h`` to ensure that we're sound in any case.

tautschnig commented:
Great, this should thus be ready to go -> @kroening

kroening commented:
To get this going, I have done some amount of cherry-picking; may I ask for a rebase, please?

martin-cs commented:
I've rebased this and am splitting it into sections; the first one is here:

https://github.com/diffblue/cbmc/pull/952

martin-cs commented:
The next part is here:

https://github.com/diffblue/cbmc/pull/961

martin-cs commented:
The next part is here:

https://github.com/diffblue/cbmc/pull/966

martin-cs commented:
Now continuing as the goto-analyzer-6 series of PRs
https://github.com/diffblue/cbmc/pull/1475
https://github.com/diffblue/cbmc/pull/1487
